### PR TITLE
[Review] URLs for CSVs

### PR DESCRIPTION
**Pull request by quasiben**
Very much a WIP for https://github.com/rapidsai/dask-cudf/issues/109

### CODE REVIEW

**Comment by quasiben** [comment]
currently failing with the following:

```
>       df3 = dask_cudf.read_csv(f"file://{stmp_path}")

dask_cudf/io/tests/test_csv.py:23: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
dask_cudf/io/csv.py:20: in read_csv
    return func(path, **kwargs)
../dask/dask/dataframe/io/csv.py:488: in read
    **kwargs)
../dask/dask/dataframe/io/csv.py:395: in read_pandas
    path=path)
../dask/dask/dataframe/io/csv.py:203: in text_blocks_to_pandas
    categoricals = head.select_dtypes(include=['category']).columns
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <cudf.DataFrame ncols=2 nrows=720 >, key = 'select_dtypes'

    def __getattr__(self, key):
        if key != '_cols' and key in self._cols:
            return self[key]
    
>       raise AttributeError("'DataFrame' object has no attribute %r" % key)
E       AttributeError: 'DataFrame' object has no attribute 'select_dtypes'

../cudf/python/cudf/dataframe/dataframe.py:185: AttributeError
```

**Comment by quasiben** [comment]
@mrocklin do you have thoughts on abandoning now and going after read_bytes or should I make necessary changes to `cudf` ?

**Comment by mrocklin** [comment]
My hope is that the necessary change to cudf might be 

```python
class DataFrame:
    select_dtypes = pandas.select_dtypes
```

**Comment by quasiben** [comment]
@mrocklin I got things semi functional with this PR:
https://github.com/rapidsai/cudf/pull/1066 it's more than a bit hacky to make `select_dtypes` work in cudf

**Comment by quasiben** [comment]
Hmm, on second thought I think i can write a cudf native `select_dtypes` function

**Comment by quasiben** [comment]
Current status:

```python
In [7]: df = dask_cudf.read_csv('s3://dask-data/nyc-taxi/2015/yellow_tripdata_2015-01.csv', storage_options={'anon': True})                                                             

In [8]: df.head().to_pandas()                                                                                                                                                           
Out[8]: 
   VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  trip_distance  ...  mta_tax  tip_amount  tolls_amount  improvement_surcharge  total_amount\r
0         2  2015-01-15 19:05:39   2015-01-15 19:23:42                1           1.59  ...      0.5        3.25           0.0                    0.3           17.05
1         1  2015-01-10 20:33:38   2015-01-10 20:53:28                1           3.30  ...      0.5        2.00           0.0                    0.3           17.80
2         1  2015-01-10 20:33:38   2015-01-10 20:43:41                1           1.80  ...      0.5        0.00           0.0                    0.3           10.80
3         1  2015-01-10 20:33:39   2015-01-10 20:35:31                1           0.50  ...      0.5        0.00           0.0                    0.3            4.80
4         1  2015-01-10 20:33:39   2015-01-10 20:52:58                1           3.00  ...      0.5        0.00           0.0                    0.3           16.30

[5 rows x 19 columns]
```

File: dask_cudf/io/tests/test_csv.py [hunk 1]
```diff
@@ -11,11 +11,18 @@ def test_read_csv(tmp_path):
     df = dask.datasets.timeseries(dtypes={"x": int, "y": int}, freq="120s").reset_index(
         drop=True
     )
-    df.to_csv(tmp_path / "data-*.csv", index=False)
+    # tmp_path is a PosixPath -- need to convert to string
+    stmp_path = str(tmp_path / "data-*.csv")
```
Comment (mrocklin) [inline dask_cudf/io/tests/test_csv.py]: I would have thought that dask.dataframe.to_csv could handle this.  Was this failing before?  If so, can I ask you to raise an issue on Dask?

File: dask_cudf/io/csv.py [hunk 2]
```diff
@@ -8,12 +8,21 @@
...
+from dask.dataframe.io.csv import make_reader
 
 import cudf
 from libgdf_cffi import GDFError
 
 
 def read_csv(path, chunksize="256 MiB", **kwargs):
+    if "://" in path:
+        func = make_reader(cudf.read_csv, "read_csv", "CSV")
+        return func(path, **kwargs)
```
Comment (mrocklin) [inline dask_cudf/io/csv.py]: We probably want to pass chunksize through here as well

File: dask_cudf/io/tests/test_csv.py (hunk 1 above)
Comment (quasiben) [reply dask_cudf/io/tests/test_csv.py]: The issue here is not with dask.dataframe but rather the conditional check in `dask_cudf.read_csv`.  PosixPath is not iterable and thus:  `"://" in path` fails

Comment (quasiben) [reply dask_cudf/io/tests/test_csv.py]: we could str() in read_csv ?

Comment (mrocklin) [reply dask_cudf/io/tests/test_csv.py]: In that case I recommend that we convert during that check

```python
if "://" in str(path):
    ...
```

That way users can still use Path objects

Comment (mrocklin) [reply dask_cudf/io/tests/test_csv.py]: Right, that

**Comment by quasiben** [comment]
@mrocklin how do you feel about add an S3 test using a valid bucket ?  It'll increase time but will definitely validate integration

**Comment by mrocklin** [comment]
Seems reasonable to me assuming we use `pytest.importorskip('s3fs')`. Dask also already has testing infrastructure for this, you might want to look at `s3_context` in dask/bytes/tests/test_s3.py

**Comment by mrocklin** [comment]
Good to go?

```python
In [1]: import dask_cudf

In [2]: df = dask_cudf.read_csv('gcs://anaconda-public-data/nyc-taxi/csv/2015/yellow_*.csv')

In [3]: df
Out[3]: <dask_cudf.DataFrame | 273 tasks | 91 npartitions>

In [4]: df.head()
Out[4]: <cudf.DataFrame ncols=19 nrows=5 >

In [5]: _.to_pandas()
Out[5]:
   VendorID tpep_pickup_datetime  ... improvement_surcharge  total_amount
0         2  2015-01-15 19:05:39  ...                   0.3         17.05
1         1  2015-01-10 20:33:38  ...                   0.3         17.80
2         1  2015-01-10 20:33:38  ...                   0.3         10.80
3         1  2015-01-10 20:33:39  ...                   0.3          4.80
4         1  2015-01-10 20:33:39  ...                   0.3         16.30

[5 rows x 19 columns]

In [6]: df.dtypes
Out[6]:
VendorID                          int64
tpep_pickup_datetime     datetime64[ms]
tpep_dropoff_datetime    datetime64[ms]
passenger_count                   int64
trip_distance                   float64
pickup_longitude                float64
pickup_latitude                 float64
RateCodeID                        int64
store_and_fwd_flag                int32
dropoff_longitude               float64
dropoff_latitude                float64
payment_type                      int64
fare_amount                     float64
extra                           float64
mta_tax                         float64
tip_amount                      float64
tolls_amount                    float64
improvement_surcharge           float64
total_amount                    float64
dtype: object
```

**Comment by quasiben** [comment]
Adding in the s3 test

**Pull request merged**